Refresh architecture description and diagrams.

jhkrug opened the pull request:
First commit, long way to go.

I'd like to work though this with other people before doing too much more. Hence, the early draft PR.

There is TODO: comments(s) in the markdown to help me work through this.

When complete, this will close #258.

jvanz commented on docs/explanations/architecture.md:
```diff
@@ -72,6 +76,49 @@ On a new cluster, the Kubewarden components defined are:
 - the `kubewarden-controller` Deployment
 - a `PolicyServer` Custom Resource named `default`.
 
+<!--TODO: Need to work through this a bit, before progressing to the next
+diagram. Is this OK, how does the text above match with the diagram below. The
+two diagrams are quite different, is anything being lost? Gained? In the
+original, the lines around the planes is dashed. What does that signify? In the
+original there is separation between The control palnes and the K8s API server.
+What are they there for? What are they imparting? There is no further mention
+of them in the text. They are never shown connecting to anything else. If they
+can be removed it may simplify layout as everything becomes more complicated. -->
```
I would say we need to rework the diagram in this page. I know what they trying to show. They try to help demonstrate visually what is written. But they are failing on that.  I'm not sure why. Maybe lack of legend? Maybe wrong diagram type? I'm not sure. But I think if they are not helping, we should remove them. 

We can remove the diagrams for now, improve the understanding using only text and re add the diagram again if necessary.

jhkrug commented:
Hi @jvanz . Getting this going again. I've created architecture-2.md. I've removed the diagrams. At the start of the text, I've added some points that should perhaps be included in the page. When we have a better grasp on the text components, we can look at diagramming again.

Can you take a look please?

jvanz commented on docs/explanations/architecture-2.md:
```diff
@@ -0,0 +1,239 @@
+[specification reference](../reference/spec/01-intro-spec.md)
+section of the documentation.
+
+The `policy-server` exits with an error
+if one or more policies received wrong configuration parameters
+from the policy specification provided by the user.
```
This is not true anymore. In a recent change, the policy server has been changed to start even if with a invalid policy. What happens now is that all the admission request sent to be evaluated by the invalid policy will return an error.

jvanz commented on docs/explanations/architecture-2.md:
```diff
@@ -0,0 +1,239 @@
+based on the endpoint that received the request,
+uses the correct policy to evaluate it.
+
+Each policy is evaluated inside its own dedicated WebAssembly sandbox.
+The communication between `policy-server` (the "host")
+and the WebAssembly policy (the "guest")
+uses the waPC communication protocol.
+The protocol description is part of the [writing policies](../tutorials/writing-policies/index.md) documentation.
```
Maybe is worth to mention the [WASI](https://docs.kubewarden.io/tutorials/writing-policies/wasi/intro-wasi) policies as well

jvanz commented on docs/explanations/architecture-2.md:
```diff
@@ -0,0 +1,239 @@
+A `PolicyServer` resource defines each `policy-server`
+and a `ClusterAdmissionPolicy` or `AdmissionPolicy` resource defines each policy.
+
+A `ClusterAdmissionPolicy` is bound to a `PolicyServer`.
```
`AdmissionPolicy` also needs to be bound to a `PolicyServer`

jvanz commented on docs/explanations/architecture-2.md:
```diff
@@ -0,0 +1,239 @@
+
+### Reconciliation of `policy-server`
+
+When creating, modifying or deleting a `ClusterAdmissionPolicy`,
```
```suggestion
When creating, modifying or deleting a `ClusterAdmissionPolicy` or `AdmissionPolicy`,
```

jvanz commented on docs/explanations/architecture-2.md:
```diff
@@ -0,0 +1,239 @@
+
+## What is a policy
+
+A policy is ???????. Sample as YAML?
```
A policy is a OCI artifact that contains a webassembly module (the policy code) and the metadata required by the PolicyServer to perform admission request validations/mutations.

jvanz commented on docs/explanations/architecture-2.md:
```diff
@@ -0,0 +1,331 @@
+It uses policies written in a programming language of your choosing.
+This language must generate a WebAssembly binary for Kubewarden to use.
+
+## Something to clear up?
+
+What's the difference between
+
+- PolicyServer and policy-server.
+
+An up-front explanation of this type of usage would be useful. I think it
+occurs elsewhere in the docs, with other terms as well. Does it need
+clarification?
```
As far as I can see, we use both term interchangeably. They usually refers to the policy engine running in the cluster. But I thinks it's a good idea to create some standard for this term.

jvanz commented on docs/explanations/architecture-2.md:
```diff
@@ -0,0 +1,331 @@
+
+When setup by the Kubewarden controller,
+the policy server receives admission requests directly from the Kubernetes control plane,
+using `ValidationWebhooks`.
```
```suggestion
using `ValidationWebhooks` and `MutatingWebhooks`.
```

jvanz commented on docs/explanations/architecture-2.md:
```diff
@@ -0,0 +1,331 @@
+the `policy-server` Deployment is ready to evaluate an
+[`AdmissionReview`](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#webhook-request-and-response).
+
+Once Kubewarden marks the `policy-server` Deployment as `Ready`,
+the `kubewarden-controller` makes the Kubernetes API server
+aware of the new policy by creating either a
+`MutatingWebhookConfiguration` or a `ValidatingWebhookConfiguration` object.
```
Actually, the controller will create/update the webhooks when the policy is "uniquely reachable". Which means that all the policy server instances running in the cluster have the latest policy configuration running. That's necessary because due the rollout of of the policy servers works,  during the deploy of a policy is possible to have the same policy (one on each policy server) with two different configuration.

jhkrug replied:
I had a go, we can review in the next iteration.

jvanz reviewed (commented):
The new arch document is getting better! Thanks for that! 

Besides the comments that I've left, the text os nice and is describing what Kubewarden components do. 

About the diagrams, I'm not sure how to describe that. I'll think about something and share with you later.

viccuad commented on docs/explanations/architecture-2.md:
```diff
@@ -67,7 +69,7 @@ ensuring that security and compliance policy enforcement is consistent across th
 
 When setup by the Kubewarden controller,
-the policy server receives admission requests directly from the Kubernetes control plane,
+the PolicyServer server receives admission requests directly from the Kubernetes control plane,
```
```suggestion
the policy-server server receives admission requests directly from the Kubernetes control plane,
```

viccuad commented on docs/explanations/architecture-2.md:
```diff
@@ -130,8 +132,8 @@ The Kubewarden consists of these components:
 
-- [Policy server](https://github.com/kubewarden/policy-server):
-  The policy server receives requests for validation.
+- [PolicyServer](https://github.com/kubewarden/policy-server):
```
```suggestion
- [Policy server](https://github.com/kubewarden/policy-server):
```

Oof this is difficult. This refers to the server, a policy-server, which receives requests. It isn't the CR PolicyServer.

viccuad commented on docs/explanations/architecture-2.md:
```diff
@@ -130,8 +132,8 @@ The Kubewarden consists of these components:
-- [Policy server](https://github.com/kubewarden/policy-server):
-  The policy server receives requests for validation.
+- [PolicyServer](https://github.com/kubewarden/policy-server):
+  The PolicyServer receives requests for validation.
```
```suggestion
  The policy-server receives requests for validation.
```

viccuad reviewed (commented):
Looks good to me! Time to move architecture-2.md into architecture.md?

fabriziosestito reviewed (approved):
LGTM

flavio reviewed (approved):
@jhkrug looks good. Thanks for the improvements!
